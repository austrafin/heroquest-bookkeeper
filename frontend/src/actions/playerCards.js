export const ADD = "playerCards/ADD";
export const ADD_AFTER = "playerCards/ADD_AFTER";
export const DELETE = "playerCards/DELETE";
export const DELETE_AFTER = "playerCards/DELETE_AFTER";
export const UPDATE_BASE_VALUES = "playerCards/UPDATE_BASE_VALUES";
export const LOAD = "playerCards/LOAD";
export const CARDS_LOADED = "playerCards/CARDS_LOADED";
export const SET_VALUE = "playerCards/SET_VALUE";
export const INCREMENT = "playerCards/INCREMENT";
export const DECREMENT = "playerCards/DECREMENT";
export const INITIALISE = "playerCards/INITIALISE";
export const SET_SELECTED_IMAGE = "playerCards/SET_SELECTED_IMAGE";
export const UPLOAD_IMAGE = "playerCards/UPLOAD_IMAGE";
export const ADD_ARMORY_ITEM = "playerCards/ADD_ARMORY_ITEM";
export const DELETE_ARMORY_ITEM = "playerCards/DELETE_ARMORY_ITEM";
export const CLEAR_PENDING_CHANGES = "playerCards/CLEAR_PENDING_CHANGES";
export const SET_INPUT_VALUE = "playerCards/SET_INPUT_VALUE";

export function addPlayerCard(values) {
  return {
    type: ADD,
    values: values,
  };
}

export function addPlayerCardPostAction() {
  return { type: ADD_AFTER };
}

export function deletePlayerCard(cardId) {
  return {
    type: DELETE,
    cardId: cardId,
  };
}

export function deletePlayerCardPostAction(cardId) {
  return {
    type: DELETE_AFTER,
    cardId: cardId,
  };
}

export function updateBaseValues(values, cardId) {
  return {
    type: UPDATE_BASE_VALUES,
    cardId: cardId,
    values: values,
  };
}

export function loadPlayerCards() {
  return {
    type: LOAD,
  };
}

export function setValue(value, label, cardId) {
  return {
    type: SET_VALUE,
    value: value,
    label: label,
    cardId: cardId,
  };
}

export function increment(value, label, cardId) {
  return {
    type: INCREMENT,
    incrementValue: value,
    label: label,
    cardId: cardId,
  };
}

export function decrement(value, label, cardId) {
  return {
    type: DECREMENT,
    incrementValue: value,
    label: label,
    cardId: cardId,
  };
}

export function setSelectedImageFile(selectedFile, cardId) {
  return {
    type: SET_SELECTED_IMAGE,
    selectedFile: selectedFile,
    cardId: cardId,
  };
}

export function uploadImage(selectedFile, cardId) {
  return {
    type: UPLOAD_IMAGE,
    selectedFile: selectedFile,
    cardId: cardId,
  };
}

export function addArmoryItem(cardId, itemId) {
  return {
    type: ADD_ARMORY_ITEM,
    cardId: cardId,
    itemId: itemId,
  };
}

export function deleteArmoryItem(cardId, itemId) {
  return {
    type: DELETE_ARMORY_ITEM,
    cardId: cardId,
    itemId: itemId,
  };
}

export function setInputValue(value, status, cardId) {
  return {
    type: SET_INPUT_VALUE,
    status: status,
    value: value,
    cardId: cardId,
  };
}

export function initialisePlayerCards(initialValues) {
  return { type: INITIALISE, data: initialValues };
}

export function setCardsLoaded(value) {
  return { type: CARDS_LOADED, value: value };
}

export function clearPendingChanges() {
  return { type: CLEAR_PENDING_CHANGES };
}
