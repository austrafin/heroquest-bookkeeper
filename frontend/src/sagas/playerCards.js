import { takeLatest, put, delay } from "redux-saga/effects";
import {
  ADD,
  DELETE,
  LOAD,
  INCREMENT,
  DECREMENT,
  UPLOAD_IMAGE,
  ADD_ARMORY_ITEM,
  DELETE_ARMORY_ITEM,
  UPDATE_BASE_VALUES,
  initialisePlayerCards,
  addPlayerCardPostAction,
  deletePlayerCardPostAction,
  setCardsLoaded,
  setSelectedImageFile,
  clearPendingChanges,
} from "../actions/playerCards";
import {
  REDUX_STORE_FIELDS as Constants,
  DB_FIELDS as DB,
  ALT_IMAGE_PATH,
} from "../constants/player_card.constants";
import axios from "axios";
import store from "../store";

function mapBaseValuesReduxToDB(values) {
  return {
    [DB.CHARACTER_NAME]: values[Constants.CHARACTER_NAME],
    [DB.BASE_BODY_POINTS]: values[Constants.BASE_BODY_POINTS],
    [DB.BASE_MIND_POINTS]: values[Constants.BASE_MIND_POINTS],
    [DB.BASE_DIAGONAL_POINTS]: values[Constants.BASE_DIAGONAL_POINTS],
    [DB.BASE_MELEE_POINTS]: values[Constants.BASE_MELEE_POINTS],
    [DB.BASE_RANGED_POINTS]: values[Constants.BASE_RANGED_POINTS],
    [DB.BASE_DEFENCE_POINTS]: values[Constants.BASE_DEFENCE_POINTS],
    [DB.BASE_MOVEMENT_POINTS]: values[Constants.BASE_MOVEMENT_POINTS],
  };
}

export function* loadPlayerCardData() {
  yield put(setCardsLoaded(false));
  const initialValues = {};

  yield axios
    .get(process.env.REACT_APP_API_BASE_URL + "player_cards")
    .then((response) => {
      if (response.status === 200) {
        response.data.forEach((card) => {
          const values = {};
          values[Constants.CHARACTER_NAME] = card[DB.CHARACTER_NAME];
          values[Constants.BASE_BODY_POINTS] = card[DB.BASE_BODY_POINTS];
          values[Constants.BODY_POINTS] = card[DB.BODY_POINTS];
          values[Constants.BASE_MIND_POINTS] = card[DB.BASE_MIND_POINTS];
          values[Constants.MIND_POINTS] = card[DB.MIND_POINTS];
          values[Constants.BASE_MELEE_POINTS] = card[DB.BASE_MELEE_POINTS];
          values[Constants.MELEE_POINTS] = card[DB.BASE_MELEE_POINTS];
          values[Constants.BASE_RANGED_POINTS] = card[DB.BASE_RANGED_POINTS];
          values[Constants.RANGED_POINTS] = card[DB.BASE_RANGED_POINTS];
          values[Constants.BASE_DIAGONAL_POINTS] =
            card[DB.BASE_DIAGONAL_POINTS];
          values[Constants.DIAGONAL_POINTS] = card[DB.BASE_DIAGONAL_POINTS];
          values[Constants.BASE_DEFENCE_POINTS] = card[DB.BASE_DEFENCE_POINTS];
          values[Constants.DEFENCE_POINTS] = card[DB.BASE_DEFENCE_POINTS];
          values[Constants.BASE_MOVEMENT_POINTS] =
            card[DB.BASE_MOVEMENT_POINTS];
          values[Constants.MOVEMENT_POINTS] = card[DB.BASE_MOVEMENT_POINTS];
          values[Constants.GOLD] = card[DB.GOLD];
          values[Constants.ARMORY_ITEMS] = card[DB.ARMORY_ITEMS];

          if (card[DB.IMAGE] !== undefined) {
            const base64 = btoa(
              new Uint8Array(card[DB.IMAGE].data).reduce(
                (data, byte) => data + String.fromCharCode(byte),
                ""
              )
            );

            values[Constants.IMAGE] = "data:image/png;base64," + base64;
          } else {
            values[Constants.IMAGE] = ALT_IMAGE_PATH;
          }

          initialValues[card[DB.ID]] = values;
        });
      }
    })
    .catch((error) => {
      console.log(error);
    });
  yield put(initialisePlayerCards(initialValues));
  yield put(setCardsLoaded(true));
}

export function* addPlayerCard(action) {
  yield put(setCardsLoaded(false));
  yield axios
    .post(
      process.env.REACT_APP_API_BASE_URL + "player_cards/add",
      mapBaseValuesReduxToDB(action.values)
    )
    .catch((error) => {
      console.log(error);
    });
  yield put(addPlayerCardPostAction());
  yield put(setCardsLoaded(true));
}

export function* deletePlayerCard(action) {
  yield put(setCardsLoaded(false));
  yield axios
    .delete(
      process.env.REACT_APP_API_BASE_URL + "player_cards/" + action.cardId
    )
    .catch((error) => {
      console.log(error);
    });
  yield put(deletePlayerCardPostAction(action.cardId));
  yield put(setCardsLoaded(true));
}

export function* addArmoryItem(action) {
  const body = { itemId: action.itemId };
  try {
    let responseStatus = null;

    yield axios
      .post(
        process.env.REACT_APP_API_BASE_URL +
          "player_cards/add_armory_item/" +
          action.cardId,
        body
      )
      .then((response) => {
        responseStatus = response.status;
      });

    if (responseStatus === 200) {
      yield put({ type: LOAD });
    }
  } catch (error) {
    console.log(error);
  }
}

export function* deleteArmoryItem(action) {
  const body = { itemId: action.itemId };
  try {
    let responseStatus = null;

    yield axios
      .patch(
        process.env.REACT_APP_API_BASE_URL +
          "player_cards/delete_armory_item/" +
          action.cardId,
        body
      )
      .then((response) => {
        responseStatus = response.status;
      });

    if (responseStatus === 200) {
      yield put({ type: LOAD });
    }
  } catch (error) {
    console.log(error);
  }
}

export function* updateDatabase() {
  yield delay(200);

  const pendingChanges = store.getState().playerCards.pendingChanges;
  const updateValues = {};

  Object.entries(pendingChanges).map(([key, value]) => {
    updateValues[key] = {};

    if (Constants.BODY_POINTS in pendingChanges[key])
      updateValues[key][DB.BODY_POINTS] =
        pendingChanges[key][Constants.BODY_POINTS];
    if (Constants.MIND_POINTS in pendingChanges[key])
      updateValues[key][DB.MIND_POINTS] =
        pendingChanges[key][Constants.MIND_POINTS];
    if (Constants.GOLD in pendingChanges[key])
      updateValues[key][DB.GOLD] = pendingChanges[key][Constants.GOLD];
    return updateValues;
  });

  yield axios
    .post(
      process.env.REACT_APP_API_BASE_URL + "player_cards/update",
      updateValues
    )
    .catch((error) => {
      console.log(error);
    });
  yield put(clearPendingChanges());
}

export function* updateBaseValues(action) {
  yield delay(200);
  yield axios
    .post(process.env.REACT_APP_API_BASE_URL + "player_cards/update", {
      [action.cardId]: mapBaseValuesReduxToDB(action.values),
    })
    .catch((error) => {
      console.log(error);
    });
  yield put(setCardsLoaded(false));
}

export function* uploadImage(action) {
  if (action.selectedFile !== null) {
    const formData = new FormData();
    formData.append(
      "characterImage",
      action.selectedFile,
      action.selectedFile.name
    );
    try {
      let responseStatus = null;

      yield axios
        .post(
          process.env.REACT_APP_API_BASE_URL +
            "player_cards/upload_image/" +
            action.cardId,
          formData
        )
        .then((response) => {
          responseStatus = response.status;
        });
      if (responseStatus === 200) {
        yield put(setSelectedImageFile(null, action.cardId));
      }
    } catch (error) {
      console.log(error);
    }
  }
}

export const playerCardsSagas = [
  takeLatest(ADD, addPlayerCard),
  takeLatest(LOAD, loadPlayerCardData),
  takeLatest(DELETE, deletePlayerCard),
  takeLatest(INCREMENT, updateDatabase),
  takeLatest(DECREMENT, updateDatabase),
  takeLatest(UPDATE_BASE_VALUES, updateBaseValues),
  takeLatest(UPLOAD_IMAGE, uploadImage),
  takeLatest(ADD_ARMORY_ITEM, addArmoryItem),
  takeLatest(DELETE_ARMORY_ITEM, deleteArmoryItem),
];
