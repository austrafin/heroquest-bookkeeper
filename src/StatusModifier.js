import React, { useRef } from "react";
import NumberInput from "./NumberInput";
import { Button, ButtonGroup, Grid } from "@material-ui/core";

function addToStatus(input) {
  input.current.value = "";
}

function takeFromStatus(input) {
  input.current.value = "";
}

const StatusModifier = ({ maxValue = 100, step = 1, labelText }) => {
  const inputRef = useRef(null);

  return (
    <Grid container spacing={1}>
      <Grid item xs>
        <NumberInput
          reference={inputRef}
          labelText={labelText}
          maxValue={maxValue}
          step={step}
        />
      </Grid>
      <Grid item xs>
        <ButtonGroup color="primary" variant="contained" fullWidth={true}>
          <Button onClick={() => addToStatus(inputRef)}>+</Button>
          <Button onClick={() => takeFromStatus(inputRef)}>-</Button>
        </ButtonGroup>
      </Grid>
    </Grid>
  );
};

export default StatusModifier;
