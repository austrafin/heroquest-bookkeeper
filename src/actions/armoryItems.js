export const SET = "armoryItems/SET";
export const UPDATE = "armoryItems/UPDATE";

export function setArmoryItem(value, key, id) {
  return {
    type: SET,
    key: key,
    id: id,
    value: value,
  };
}

export function updateArmoryItem(id) {
  return {
    type: UPDATE,
    id: id,
  };
}
