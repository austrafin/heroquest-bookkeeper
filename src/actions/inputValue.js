export const SET = "inputValue/SET";

export function setInputValue(value, status) {
  return {
    type: SET,
    status: status,
    value: value
  };
}
