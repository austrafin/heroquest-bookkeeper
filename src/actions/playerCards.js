export const ADD = "playerCards/ADD";
export const ADD_AFTER = "playerCards/ADD_AFTER";
export const LOAD = "playerCards/LOAD";
export const CARDS_LOADED = "playerCards/CARDS_LOADED";

export function addPlayerCard(values) {
  return {
    type: ADD,
    values: values,
  };
}

export function loadPlayerCards() {
  return {
    type: LOAD,
  };
}
