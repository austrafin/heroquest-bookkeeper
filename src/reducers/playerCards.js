import {
  ADD,
  ADD_AFTER,
  CARDS_LOADED,
  SET_VALUE,
  INCREMENT,
  DECREMENT,
  INITIALISE,
  SET_SELECTED_IMAGE,
  UPDATE_BASE_VALUES,
} from "../actions/playerCards";

function difference(state, action, decrement) {
  let incrementValue = action.incrementValue;

  if (decrement) {
    incrementValue *= -1;
  }

  if (action.cardId in state.cardData) {
    return {
      ...state,
      cardData: {
        ...state.cardData,
        [action.cardId]: {
          ...state.cardData[action.cardId],
          [action.label]:
            state.cardData[action.cardId][action.label] + incrementValue,
        },
      },
    };
  }
  return {
    ...state,
    cardData: {
      ...state.cardData,
      [action.cardId]: {
        ...state.cardData[action.cardId],
        [action.label]: 0 + incrementValue,
      },
    },
  };
}

export default (state = {}, action) => {
  switch (action.type) {
    case SET_VALUE:
      if (action.cardId in state.cardData) {
        return {
          ...state,
          cardData: {
            ...state.cardData,
            [action.cardId]: {
              ...state.cardData[action.cardId],
              [action.label]: action.value,
            },
          },
        };
      }
      return state;
    case INCREMENT:
      return difference(state, action, false);
    case DECREMENT:
      return difference(state, action, true);
    case INITIALISE:
      return {
        ...state,
        cardData: {
          ...action.data,
        },
      };
    case ADD:
      return { ...state, newCardUploading: true };
    case ADD_AFTER:
      return { ...state, newCardUploading: false };
    case CARDS_LOADED:
      return { ...state, cardsLoaded: action.value };
    case SET_SELECTED_IMAGE:
      return {
        ...state,
        cardData: {
          ...state.cardData,
          [action.cardId]: {
            ...state.cardData[action.cardId],
            selectedImageFile: action.selectedFile,
          },
        },
      };
    case UPDATE_BASE_VALUES:
      console.log(action);
      return {
        ...state,
        cardData: {
          ...state.cardData,
          [action.cardId]: {
            ...state.cardData[action.cardId],
            baseBodyPoints: action.values.baseBodyPoints,
            baseMindPoints: action.values.baseMindPoints,
            baseMeleePoints: action.values.baseMeleePoints,
            baseRangedPoints: action.values.baseRangedPoints,
            baseDiagonalPoints: action.values.baseDiagonalPoints,
            baseDefencePoints: action.values.baseDefencePoints,
            baseMovementPoints: action.values.baseMovementPoints,
          },
        },
      };
    default:
      return state;
  }
};
