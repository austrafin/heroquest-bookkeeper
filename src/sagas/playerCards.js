import { takeLatest, put } from "redux-saga/effects";
import { ADD, ADD_AFTER } from "../actions/playerCards";
import axios from "axios";

function* addPlayerCard(action) {
  yield axios
    .post("http://localhost:5000/player_cards/add", action.values)
    .catch((error) => {
      console.log(error);
    });
  yield put({ type: ADD_AFTER });
}

export const playerCardsSagas = [takeLatest(ADD, addPlayerCard)];
