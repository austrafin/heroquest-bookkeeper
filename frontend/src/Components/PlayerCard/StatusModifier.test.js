import React from "react";
import { mount } from "enzyme";
import StatusModifier from "./StatusModifier";
import { Provider } from "react-redux";
import { increment, decrement } from "../../actions/playerCards";
import configureStore from "redux-mock-store";

jest.mock("../../actions/playerCards");

describe("StatusModifier", () => {
  const mockStore = configureStore([]);
  const cardId = "test_123";
  const store = mockStore({
    playerCards: {
      cardData: {
        [cardId]: { inputValues: {} },
      },
    },
  });

  store.dispatch = jest.fn();

  beforeEach(() => {
    increment.mockClear();
    decrement.mockClear();
  });

  const component = mount(
    <Provider store={store}>
      <StatusModifier
        labelText={"test label"}
        maxValue={100}
        step={1}
        defaultValue={1}
        labelParameter="testParameter"
        cardId={cardId}
      />
    </Provider>
  );

  it("renders correctly", () => {
    expect(component).toMatchSnapshot();
  });

  it("checks that the increment action is dispatched when the increment button is clicked.", () => {
    component.find('[data-test="increment"]').find("button").simulate("click");
    expect(increment).toHaveBeenCalledTimes(1);
  });

  it("checks that the decrement action is dispatched when the decrement button is clicked.", () => {
    component.find('[data-test="decrement"]').find("button").simulate("click");
    expect(decrement).toHaveBeenCalledTimes(1);
  });
});
