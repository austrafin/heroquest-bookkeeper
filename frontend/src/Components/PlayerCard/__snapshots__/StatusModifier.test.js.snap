// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusModifier renders correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
  >
    <div
      className="MuiFormControl-root MuiTextField-root"
      data-test="test-number-input"
      onKeyPress={[Function]}
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink={true}
      >
        test label
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          defaultValue={1}
          disabled={false}
          max={100}
          min={1}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          step={1}
          type="number"
        />
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
  >
    <div
      className="MuiButtonGroup-root MuiButtonGroup-fullWidth MuiButtonGroup-contained"
      role="group"
      style={
        Object {
          "marginTop": 12,
        }
      }
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-fullWidth"
        data-test="increment"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          +
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-fullWidth"
        data-test="decrement"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          -
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
